Queued dxDrawPrimitive and dxDrawMaterialPrimitive

forkerer opened the pull request:
It's pretty much [this](https://github.com/multitheftauto/mtasa-blue/pull/301) pull request, but with proper queueing, state saving, and ability to be called from anywhere, not only onClientRender event.
Also there's added postGUI parameter:

`bool dxDrawPrimitive( string primitiveType, [ postGUI = true ] table vertex1, table vertex2, table vertex3, ... )`

`bool dxDrawMaterialPrimitive( string primitiveType, material, [ postGUI = true ] table vertex1, table vertex2, table vertex3, ... )`

Commit:
Merge remote-tracking branch 'multitheftauto/master' into dxDrawPrimitive

Commit:
Auto stash before merge of "dxDrawPrimitive" and "multitheftauto/master"

Basic dxDrawPrimitive

Commit:
Added material caching, so no settings are changed needlessly

Commit:
Merge remote-tracking branch 'multitheftauto/master' into dxDrawPrimitive

Commit:
Merge remote-tracking branch 'multitheftauto/master' into dxDrawPrimitive

Commit:
Small fixes

CrosRoad95 commented:
HE STOLE MY PULL REQUEST!

forkerer commented:
yes

Commit:
Naming changes, formatting

4O4 commented:
While I really appreciate the work you did, I have to say the way you did it feels really wrong. This should be kept in scope of one pull request - you could ask @CrosRoad95 for write permissions to his fork or send him a patch to apply manually. But this is the tiny issue here. The bigger one is the legal one - code might be open source but there's a thing called copyright. You two should have a short conversation and come to some agreement :)

forkerer commented:
Ye, i completely agree that it isn't the best way of doing it. My motivation for that was that i was telling CrosRoad exactly what's wrong with his dxDrawPrimitive since he created the pull request, as it is just flawed. I wont mind if this request is closed and not accepted as long as the original one is modified to work properly. It's not an attack on him, i just want the primitive drawing to be added, and this is the easiest way for me to show how it could be done.

CrosRoad95 commented:
i can close my pr and you can marge this one.
also i noticed that your function don't have any limit of vertices, i tested and limit 4096 vertices should be ok

CrosRoad95 commented:
i have one question: should be option to rotate and scale primitives or not? just like dxDrawText support this

forkerer commented:
Would be pretty complicated to add right now i think, it's a good idea tho. Additional rotate/scale center parameter sounds nice.

Commit:
Fixed conditions checking for Pointlist, cleaned up classes a bit

- dxDrawPrimitive doesn't require you to draw at least 2 points now
- Batcher classes did draw only a single point at all times anyways tho, fixed that
- Cleaned up CPrimitiveBatcher.cpp
- Replaced NULLs with nullptrs

Commit:
Fix for pointlist checking

Noticed that with last commit it would be possible to try draw 0 points without error

Commit:
linestrip condition checking

qaisjp commented:
#301 has more gifs and so is better /close (thanks for your hard work @CrosRoad95)

ccw808 commented:
sam1ler, the code you commented on has been copied from CTileBatcher.cpp, so I don't think your suggestions should be done in isolation.

forkerer commented on Client/core/Graphics/CPrimitiveBatcher.cpp:
```diff
@@ -0,0 +1,214 @@
+    case D3DPT_LINESTRIP:
+        iSize = iCollectionSize - 1;
+        break;
+    case D3DPT_TRIANGLEFAN:
```
D3DPT_TRIANGLEFANsize needs to be same as D3DPT_TRIANGLESTRIP, so break is unnecessary, it uses shared case

ghost commented:
@ccw808 I see, I didn't know that. This is a lot of code. We have a problem of redundancy here. How about sharing the code between CTileBatcher.cpp, CPrimitiveMaterialBatcher.cpp, and CPrimitiveBatcher.cpp? I can refactor the code in CTileBatcher.cpp, then the author of PR can make appropriate changes to the PR.

ccw808 commented:
@sam1ler it's up to you. (I would put refactor on the TODO list and leave it for now)

qaisjp commented:
When merging (after 1.5.6), please squash and add the following line to the end of the commit message:

```
Co-authored-by: CrosRoad95 <sebajura1234@gmail.com>
```

Commit:
Merge remote-tracking branch 'multitheftauto/master' into dxDrawPrimitive

Commit:
hopefully no more bad stuff

Commit:
gitignore again
